app/lib/account/backend.dart

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:googleapis/oauth2/v2.dart' as oauth2_v2;
+import 'package:http/http.dart' as http;
+import 'package:logging/logging.dart';
+import 'package:pool/pool.dart';
+import 'package:retry/retry.dart';
+
+import 'models.dart';
+
+final _logger = new Logger('pub.account.backend');
+final _retry = RetryOptions();
+
+/// The pub client's OAuth2 identifier.
+final _pubAudience = '818368855108-8grd2eg9tj9f38os6f1urbcvsq399u8n.apps.'
+    'googleusercontent.com';
+
+/// Sets the account backend service.
+void registerAccountBackend(AccountBackend backend) =>
+    ss.register(#_accountBackend, backend);
+
+/// The active account backend service.
+AccountBackend get accountBackend =>
+    ss.lookup(#_accountBackend) as AccountBackend;
+
+/// Represents the backend for the account handling and authentication.
+class AccountBackend {
+  final _pool = Pool(1);
+  final DatastoreDB _db;
+  final _defaultAuthProvider =
+      GoogleOauth2AuthProvider('dartlang-pub-1', _pubAudience);
+
+  AccountBackend(this._db);
+
+  Future close() async {
+    await _defaultAuthProvider.close();
+  }
+
+  Future<User> lookupOrCreateUser(String accessToken) async {
+    final auth = await _defaultAuthProvider.tryAuthenticate(accessToken);
+    if (auth == null) {
+      return null;
+    }
+    final compositeId = auth.compositeId;
+    final userKey = _db.emptyKey.append(User, id: compositeId);
+
+    final accessTime = DateTime.now().toUtc();
+    final user = await _retry.retry(() async {
+      User user = (await _db.lookup<User>([userKey])).single;
+      if (user == null) {
+        // create user
+        user = User()
+          ..parentKey = _db.emptyKey
+          ..id = compositeId
+          ..authProvider = auth.provider
+          ..authUserId = auth.userId
+          ..email = auth.email
+          ..emails = <String>[auth.email]
+          ..created = accessTime
+          ..updated = accessTime;
+        _db.commit(inserts: [user]);
+      } else {
+        // update user if e-mail has been changed
+        if (user.updateEmail(auth.email)) {
+          await _db.withTransaction((tx) async {
+            user = (await _db.lookup<User>([userKey])).single;
+            user.updateEmail(auth.email);
+            user.updated = accessTime;
+            tx.queueMutations(inserts: [user]);
+            await tx.commit();
+          });
+        }
+      }
+      return user;
+    });
+
+    if (user == null) {
+      return null;
+    }
+
+    if (!_pool.isClosed) {
```

Comment:
What does it mean that the pool is closed?
Who would ever close it?
Can we really skip the database update in that case?

Reply:
I've left out `_pool.close()` from the `AccountBackend.close()`, adding it.

It is part of the shutdown sequence (which we don't trigger right now), and at that point we shouldn't receive requests. The pool was already marked as closing, which means even if we schedule a new update, the scheduled command will not run.

However, to prevent accidental data inconsistency, we can let the code flow call `_pool.withResource`, because if the pool is already closed, it will throw an exception, making it more consistent: no access to the `User` object, no update to the `lastAccess` field.